assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java

```diff
@@ -1647,136 +1645,195 @@ public String getAssignmentStatus(String assignmentId) {
-        String assignmentReference = AssignmentReferenceReckoner.reckoner().assignment(assignment).reckon().getReference();
-        boolean allowGrade = assignment != null && allowGradeSubmission(assignmentReference);
+    public SubmissionStatus getSubmissionCanonicalStatus(AssignmentSubmission submission, boolean canGrade) {
+        SubmissionStatus status;
+        if (canGrade) {
+            status = getGradersCanonicalSubmissionStatus(submission);
+        } else {
+            status = getSubmittersCanonicalSubmissionStatus(submission);
+        }
+
+        log.debug("getSubmissionCanonicalStatus for submission {} : {}", submission, status);
+        return status;
+    }
+
+    private AssignmentConstants.SubmissionStatus getGradersCanonicalSubmissionStatus(AssignmentSubmission submission) {
+        if (submission == null) return SubmissionStatus.NO_SUBMISSION;
 
         Instant submitTime = submission.getDateSubmitted();
         Instant returnTime = submission.getDateReturned();
-        Instant lastModTime = submission.getDateModified();
-        if (submission.getSubmitted() || (!submission.getSubmitted() && allowGrade)) {
+
+        // States matching a person who can grade a submission
+        if (submission.getSubmitted()) {
             if (submitTime != null) {
                 if (submission.getReturned()) {
                     if (returnTime != null && returnTime.isBefore(submitTime)) {
                         if (!submission.getGraded()) {
-                            status = AssignmentConstants.SubmissionStatus.RESUBMITTED;
-                            if (submitTime.isAfter(assignment.getDueDate())) {
-                                status = AssignmentConstants.SubmissionStatus.LATE;
+                            if (submitTime.isAfter(submission.getAssignment().getDueDate())) {
+                                return SubmissionStatus.LATE;
+                            } else {
+                                return SubmissionStatus.RESUBMITTED;
                             }
-                        } else
-                            status = AssignmentConstants.SubmissionStatus.RETURNED;
-                    } else
-                        status = AssignmentConstants.SubmissionStatus.RETURNED;
-                } else if (submission.getGraded() && allowGrade) {
-                    status = StringUtils.isNotBlank(submission.getGrade()) ? AssignmentConstants.SubmissionStatus.GRADED : AssignmentConstants.SubmissionStatus.COMMENTED;
-                } else {
-                    if (allowGrade) {
-                        // ungraded submission
-                        status = AssignmentConstants.SubmissionStatus.UNGRADED;
+                        } else {
+                            return SubmissionStatus.RETURNED;
+                        }
                     } else {
-                        status = AssignmentConstants.SubmissionStatus.SUBMITTED;
+                        return SubmissionStatus.RETURNED;
```

Comment:
Are you positive about this line? just seems like a logic change reading through

Reply:
I think it's correct as those lines were moved to the submitters method.

side note before I modified any of this I created the unit tests on the previous code and this code passes the same tests.